fetch.bs

```diff
@@ -3225,7 +3238,112 @@ through TLS using ALPN. The protocol cannot be spoofed through HTTP requests in
 </ol>
 </div>
 
+<h3 id=https-upgrades>HTTPS upgrading</h3>
+
+<div algorithm>
+
+<p>User agents may optionally upgrade requests with <a>potentially untrustworthy URL</a>s
+to attempt to fetch them over <a>potentially trustworthy URL</a>s. If an upgraded request
+fails with a network error, it is retried over the original URL.
+
+<p>The HTTPS upgrading algorithm consists of Upgrade and Fallback algorithms, with the following
+steps:
+
+<h4 id=https-upgrades-upgrade>Upgrade algorithm</h4>
+
+<p>Given a request <var>request</var>, this algorithm will rewrite its URL if the request should be
+modified and loaded over <code>"https"</code>, via the following algorithm:
+
+<ol>
+  <li>
+    <p>If one or more of the following conditions are met, return without modifying request:
+    <ul>
+      <li><p><var>request</var> is not a navigation request whose destination is document</li>
+
+      <li><p><var>request</var>'s method is not `<code>GET</code>`</li>
+
+      <li><p><var>request</var>'s URL's scheme is not <code>"http"</code></li>
+
+      <li><p><var>request</var>'s URL's host is non-unique (i.e., contains a non-registrable or
+      non-assignable domain name such as .local or an IP address that falls in a range reserved for
+      non-publicly routable networks)</li>
+
+      <li><p>Optionally, <var>request</var>'s URL's host is exempted from upgrades by the user agent
+      </li>
+    </ul>
+  </li>
+
+  <li><p>If <var>request</var>'s <code>https-upgrade-fallback-url</code> is non-null, this is a
+  fallback request that shouldn't be upgraded again. Clear the <code>https-upgrade</code> and
+  <code>https-upgrade-fallback-url</code> flags and return <var>request</var>.</li>
+
+  <li><p>Otherwise, upgrade the request by setting the following fields:
+    <ul>
+      <li>Set <code>https-upgrade-fallback-url</code> to request's URL</li>
+      <li>Set <var>request</var>'s URL's scheme to <code>"https"</code></li>
+      <li>Set <code>https-upgrade</code> flag to true</li>
+    </ul>
+  </li>
+</ol>
+
+<h4 id=https-upgrades-fallback>Fallback algorithm</h4>
+
+<p>Runs after the Upgrade algorithm. This algorithm determines whether the upgraded
+request completed without a network error. If not, it retries the request over <code>"http"</code>.
+
+<ol>
+  <li><p>If the <code>https-upgrade</code> flag is not set, return without modifying request.</li>
+
+  <li><p>Otherwise, this is an upgraded request. Let <var>upgradedResponse</var> be the result of
+  fetching the upgraded request.</li>
+
+  <li>
+    <p>If <var>upgradedResponse</var> is a network error, initiate a fallback http load by doing the
+    following:
+
+    <ul>
+      <li><p>Set <var>request</var>'s URL to the value of <var>request</var>'s
+        <code>https-upgrade-fallback-url</code> flag</li>
+
+      <li><p>Set response to the result of running <a>HTTP-redirect fetch</a> given fetchParams and
+        <var>upgradedResponse</var>.</li>
```

Comment:
Haven't got to the service worker bits yet, but rephrased this as injecting a 307 redirect. 

Since we return a response here, I think the http fetch step that calls this algorithm needs something like this to use the return value:

`Set fallbackResponse to the result of running <a>handle an HTTPS upgrade fallback</a>. If fallbackResponse is not null, set response to fallbackResponse.`

Does that sound right?

Reply:
Service worker bits (step 3.5) and step 5 done now, I also updated the fallback algorithm definition.

The idea here is that in the steps before 6, instead of returning a network error, we return the result of `https_upgrade_fallback(request, network_error)`. With this change, Step 6 becomes the actual "Handle upgrade fallback" step. Thus, we don't need to modify the "return a network error" bits after Step 6.